typing/typecore.ml

```diff
@@ -1542,10 +1530,6 @@ and type_pat_aux
           let name = constr.cstr_name in
           raise (Error (loc, !env, Unexpected_existential (r,name, exs)))
       end;
-      (* if constructor is gadt, we must verify that the expected type has the
-         correct head *)
-      if constr.cstr_generalized then
-        unify_head_only ~refine loc env (instance expected_ty) constr;
```

Comment:
I think this is probably not safe to remove. This unification ensures that the expected type of the pattern matches the un-indexed return type of the constructor *without being allowed to add equations*. The later unification call on line 1584 has a different value of `refine` and is allowed to add equations to make things true.

I would expect this branch to allow using GADT constructors to match arbitrary abstract types. If that is the case, then we should add that as a test to the testsuite.

Reply:
Thanks for the explanation!